handler/category-dimension-import.go

```diff
@@ -64,54 +79,85 @@ func (h *CategoryDimensionImport) Handle(ctx context.Context, e *event.CategoryD
 			Code:     variable.Codes[i], // TODO can we assume this?
 			Option:   variable.Codes[i],
 			Label:    variable.Labels[i],
-		}); err != nil {
-			// TODO we might want to retry this, once retries are implemented
-			return h.instanceFailed(ctx, fmt.Errorf("error posting instance dimension option: %w", err), e)
+		}, eTag)
+		if err != nil {
+			switch errPost := err.(type) {
+			case dataset.ErrInvalidDatasetAPIResponse:
+				if errPost.Code() == http.StatusConflict {
+					_, eTag, err = h.getSubmittedInstance(ctx, e, headers.IfMatchAnyETag)
+					if err != nil {
+						return err
+					}
+					i-- // retry with new eTag value, as the instance is still in a valid state
+					continue
+				}
+			default:
+				return h.instanceFailed(ctx, fmt.Errorf("error posting instance dimension option: %w", err), e)
+			}
 		}
 	}
 
-	// Count number of options for each dimension, if all dimensions have more than 0, atomically update the state to completed
-
-	// check if this was the last dimension that was updated, according to the number of observations - then update instance and import task state to completed
-	// TODO: we can have concurrency issues with this if multiple instances of this services handle different dimensions concurrently - we need an ETag - IfMatch mechanism in dataset API
-	// it would be good enough if updateInstanceWithNewInserts returns the current value
-	if err := h.datasets.PutInstanceState(ctx, h.cfg.ServiceAuthToken, e.InstanceID, dataset.StateCompleted); err != nil {
-		return h.instanceFailed(ctx, fmt.Errorf("error updating instance to completed state: %w", err), e)
-	}
-
 	log.Info(ctx, "successfully posted all dimension options to dataset api", log.Data{
 		"codes":     variable.Codes,
 		"labels":    variable.Labels,
 		"dimension": e.DimensionID,
 	})
 
+	// Count the total number of dimension options for this instance, and update the instance state to 'completed' if all values have been imported
+	dims, eTag, err := h.datasets.GetInstanceDimensions(ctx, h.cfg.ServiceAuthToken, e.InstanceID, &dataset.QueryParams{Limit: 0}, headers.IfMatchAnyETag)
+	if err != nil {
+		return h.instanceFailed(ctx, fmt.Errorf("error counting instance dimensions: %w", err), e)
+	}
+
+	if dims.Count == resp.Dataset.Size {
+		// Check if this was the last dimension that was updated - then update instance to 'completed' state
+		// Set instance to complete if and only if all dimension options have been added and we were the last ones to add a dimension option (i.e. eTag did not change between updating and counting)
+		// This will guarantee that this is done exactly once, because if the instance changed, this request will fail with 409, and the other consumer that set the last value will do the update
+		_, err = h.datasets.PutInstanceState(ctx, h.cfg.ServiceAuthToken, e.InstanceID, dataset.StateCompleted, eTag)
+		if err != nil {
+			switch putErr := err.(type) {
+			case dataset.ErrInvalidDatasetAPIResponse:
```

Comment:
can you add some comment at the correct line in all of this to make it clear which line is seeing / dealing with the 409, so as to tie the comments above with the exact line, as its not 100% clear.

Reply:
I also added a few more comments in the loop that posts the dimension options, which has a similar error checking.